packages/wct-mocha/src/clisocket.ts

```diff
@@ -0,0 +1,164 @@
+  private readonly browserId: string;
+  constructor(browserId: string, socket: SocketIO.Socket) {
+    this.browserId = browserId;
+    this.socket = socket;
+  }
+
+  /**
+   * @param {!Mocha.Runner} runner The Mocha `Runner` to observe, reporting
+   *     interesting events back to the CLI runner.
+   */
+  observe(runner: Mocha.IRunner) {
+    this.emitEvent('browser-start', {
+      url: window.location.toString(),
+    });
+
+    // We only emit a subset of events that we care about, and follow a more
+    // general event format that is hopefully applicable to test runners beyond
+    // mocha.
+    //
+    // For all possible mocha events, see:
+    // https://github.com/visionmedia/mocha/blob/master/lib/runner.js#L36
+    runner.on('test', (test: Mocha.IRunnable) => {
+      this.emitEvent('test-start', {test: getTitles(test)});
+    });
+
+    runner.on('test end', (test: Mocha.IRunnable) => {
+      this.emitEvent('test-end', {
+        state: getState(test),
+        test: getTitles(test),
+        duration: (test as any).duration,
+        error: (test as any).err,
+      });
+    });
+
+    runner.on('fail', (test, err) => {
+      // fail the test run if we catch errors outside of a test function
+      if (test.type !== 'test') {
+        this.emitEvent(
+            'browser-fail',
+            'Error thrown outside of test function: ' + err.stack);
+      }
+    });
+
+    runner.on('childRunner start', (childRunner) => {
+      this.emitEvent('sub-suite-start', childRunner.share);
+    });
+
+    runner.on('childRunner end', (childRunner) => {
+      this.emitEvent('sub-suite-end', childRunner.share);
+    });
+
+    runner.on('end', () => {
+      this.emitEvent('browser-end');
+    });
+  }
+
+  /**
+   * @param {string} event The name of the event to fire.
+   * @param {*} data Additional data to pass with the event.
+   */
+  emitEvent(event: string, data?: any) {
+    this.socket.emit('client-event', {
+      browserId: this.browserId,
+      event: event,
+      data: data,
+    });
+  }
+
+  /**
+   * Builds a `CLISocket` if we are within a CLI-run environment; short-circuits
+   * otherwise.
+   *
+   * @param {function(*, CLISocket)} done Node-style callback.
+   */
+  static init(done: (error?: any, socket?: CLISocket) => void) {
+    const browserId = util.getParam('cli_browser_id');
+    if (!browserId) {
+      return done();
+    }
+    // Only fire up the socket for root runners.
+    if (ChildRunner.current()) {
+      return done();
+    }
+
+    util.loadScript(SOCKETIO_LIBRARY, function(error: any) {
+      if (error) {
+        return done(error);
+      }
+
+      const server = io(SOCKETIO_ENDPOINT);
+      // WTF(usergenic): The typings are super wrong or something.  The object
+      // returned by io() doesn't seem to map to the SocketIO.Server type at
+      // all.
+      const sockets = server as any as SocketIO.Namespace;  // server.sockets;
+      const errorListener = function(error?: any) {
```

Comment:
Couldn't inline this because according to latest typings for socket.io, I needed a reference to `eventListener` to call socket.off() explicitly:

```ts
      const sockets = server as any as SocketIO.Namespace;
      const errorListener = (error?: any) => {
        sockets.off('error', errorListener);
        done(error);
      };
      sockets.on('error', errorListener);
      const connectListener = () => {
        sockets.off('connect', connectListener);
        done(null, new CLISocket(browserId, sockets as any));
      };
      sockets.on('connect', connectListener);
```